fendermint/testing/materializer/src/validation.rs

```diff
@@ -0,0 +1,382 @@
+        subnet: &'a VSubnet,
+        amount: TokenAmount,
+        reference: Option<ResourceHash>,
+    ) -> anyhow::Result<()>
+    where
+        's: 'a,
+    {
+        // Debit parent balance; Credit child balance
+        self.fund_from_parent(subnet, account, amount, true)?;
+        self.ensure_unique(&subnet.parent().unwrap(), reference)?;
+        Ok(())
+    }
+
+    async fn join_subnet<'s, 'a>(
+        &'s mut self,
+        _parent_submit_config: &SubmitConfig<'a, ValidationMaterials>,
+        account: &'a VAccount,
+        subnet: &'a VSubnet,
+        collateral: Collateral,
+        balance: Balance,
+        reference: Option<ResourceHash>,
+    ) -> anyhow::Result<()>
+    where
+        's: 'a,
+    {
+        // Debit parent balance, but do not make the funds available in the child
+        self.fund_from_parent(subnet, account, collateral.0, false)?;
+        // Debit parent balance; Credit child balance
+        self.fund_from_parent(subnet, account, balance.0, true)?;
+        self.ensure_unique(&subnet.parent().unwrap(), reference)?;
+        Ok(())
+    }
+
+    async fn create_subnet_genesis<'s, 'a>(
+        &'s mut self,
+        _parent_submit_config: &SubmitConfig<'a, ValidationMaterials>,
+        _subnet: &'a VSubnet,
+    ) -> anyhow::Result<VGenesis>
+    where
+        's: 'a,
+    {
+        // We're supposed to fetch the data from the parent, there's nothing to check.
+        Ok(())
+    }
+
+    async fn create_relayer<'s, 'a>(
+        &'s mut self,
+        _parent_submit_config: &SubmitConfig<'a, ValidationMaterials>,
+        relayer_name: &RelayerName,
+        subnet: &'a VSubnet,
+        submitter: &'a VAccount,
+        _follow_node: &'a VNode,
+    ) -> anyhow::Result<VRelayer>
+    where
+        's: 'a,
+    {
+        self.ensure_contains(relayer_name)?;
+        // Check that submitter has balance on the parent.
+        let parent = parent_name(subnet)?;
+        self.ensure_balance(&parent, submitter)?;
+        Ok(relayer_name.clone())
+    }
+}
+
+/// Get the parent of a subnet, or fail if it doesn't have one.
+fn parent_name(subnet: &SubnetName) -> anyhow::Result<SubnetName> {
+    subnet
+        .parent()
+        .ok_or_else(|| anyhow!("{subnet:?} has no parent"))
+}
+
+#[cfg(test)]
+mod tests {
+
+    use crate::{manifest::Manifest, validation::validate_manifest, TestnetId};
+
+    // Unfortunately doesn't seem to work with quickcheck_async
+    // /// Run the tests with `RUST_LOG=info` to see the logs, for example:
+    // ///
+    // /// ```text
+    // /// RUST_LOG=info cargo test -p fendermint_testing_materializer prop_validation -- --nocapture
+    // /// ```
+    // fn init_log() {
+    //     let _ = env_logger::builder().is_test(true).try_init();
+    // }
+
+    /// Check that the random manifests we generate would pass validation.
+    #[quickcheck_async::tokio]
+    async fn prop_validation(id: TestnetId, manifest: Manifest) -> anyhow::Result<()> {
```

Comment:
Curious why we need a tokio executor here?

Reply:
Because the methods are asynchronous. In some other tests like the `smt_` ones I created a runtime manually, but I thought creating 100 one might be a bit of an overkill and a distraction in this test. Luckily this crate exists to make it available in quickcheck.